Pinned pytest, added datastore.reset(), fixed logs, packets, the determinism tests

kywch opened the pull request:
It'd be great if we can merge these to the main branch before moving forward with the task system, etc. Thanks!

1. pinned pytest version and added pytest.ini to allow importing testhelpers during pytest
2. added/fixed env.reset(): numpy_datastore, datastore (EntityState, ItemState), entity_manager
3. added/fixed various logs/packet: log_helper.py, map.py, entity.py, npc.py, player.py, combat.py, exchange.py, inventory.py, item.py, skill.py,
4. made the determinism and deterministic replay tests work again

kywch committed:
fixed the deterministic replay test, and pinned pytest version

kywch committed:
Merge pull request #18 from kywch/choe.kyoung-git-pr-23586-16073

determinism test, etc ...

kywch committed:
Merge branch 'CarperAI:v1.6' into choe.kyoung-git-pr-23586-16073

kywch committed:
Merge pull request #19 from kywch/choe.kyoung-git-pr-23586-16073

syncing with Joseph's push

kywch committed:
Followed pylint suggestion on list comprehension

nmmo/core/env.py:36:27: R1721: Unnecessary use of a comprehension, use list(range(1, config.PLAYER_N + 1)) instead. (unnecessary-comprehension)

daveey commented on nmmo/core/map.py:
```diff
@@ -64,7 +64,7 @@ def reset(self, map_id):
   def step(self):
     '''Evaluate updatable tiles'''
-    self.realm.log_milestone('Resource_Depleted', len(self.update_list),
+    self.realm.log_milestone('[MAP] Resource_Depleted', len(self.update_list),
```
Do we want [MAP] here? I assume the milestone name should be a string with no spaces or []

kywch replied:
done!

daveey commented on nmmo/core/realm.py:
```diff
@@ -72,6 +72,10 @@ def reset(self, map_id: int = None):
     """
     self.log_helper.reset()
+    # reset datastore tables, except the table for TileState
+    for s in [EntityState, ItemState]:
```
What do you think about doing:

EntityState.clear_datastore()
ItemState.clear_datastore()

This would avoid protected-access calls.

Also, we could put this into players.reset() but I'm fine with it here

kywch replied:
The relationship between SerializedState (like EntityState, ItemState), DataStore, and DataTable was hard to understand for me. EntityState and ItemState don't seem to hold actual data, and I needed to provide `object_type: str`, which is `_name` of SerializedState, to DataStore to get access to the numpy tables.

Is it necessary to keep the name of SerializedState private?

kywch replied:
I see that players.reset() does delete player entries in the numpy tables. But I also saw that the id allocators were not reset, so the deterministic test failed (the ids were not the same after resetting the env).

daveey replied:
clear_datastore() would be a new method you would define on the State object in the subclass() method.

I think it might actually be better to reset the datastore in the determinism test. There isn't any reason to reset the datastore in realm, maybe we can just assert that the datastore is empty during reset().

daveey commented on nmmo/systems/combat.py:
```diff
@@ -92,9 +92,9 @@ def attack(realm, player, target, skillFn):
 
     if player.is_player:
-        realm.log_milestone(f'Damage_{skill_name}', damage,
+        realm.log_milestone(f'[PlayerID: {player.ent_id}] Damage_{skill_name}', damage,
```
Again, I feel like the first arg to log_milestone should be a milestone name. It looks like you might want to add tags={"PlayerID": player.ent_id} which seems like a good idea

kywch replied:
Just added tags to realm.py, Realm.log_milestone(), and used it through out!

daveey commented on tests/test_determinism.py:
```diff
@@ -1,337 +1,76 @@
+    cls.config = ScriptedAgentTestConfig()
+    env = ScriptedAgentTestEnv(cls.config)
+
+    print('[TestDeterminism] Setting up the reference env with seed', cls.rand_seed)
```
how about log.info instead of prints?

can you also add a pylint check for disallowing prints?

kywch replied:
Made pylint to disallow print and replaced print with logging.info() or logging.error().

daveey commented on tests/test_determinism.py:
```diff
@@ -1,337 +1,76 @@
+    print('[TestDeterminism] Setting up the reference env with seed', cls.rand_seed)
+    cls.init_obs_src = env.reset(seed=cls.rand_seed)
+    cls.actions_src = []
+    print('Running', cls.horizon, 'tikcs')
```
is tikcs a typo?

kywch replied:
Fixed!

daveey commented on tests/test_determinism.py:
```diff
@@ -1,337 +1,76 @@
+    cls.final_npcs_rep = npcs_rep
+
+  def test_func_are_observations_equal(self):
+    self.assertTrue(are_observations_equal(self.init_obs_src, self.init_obs_src))
```
nitpick, but how about observations_are_equal()

kywch replied:
Changed!

daveey commented on tests/test_determinism.py:
```diff
@@ -1,337 +1,76 @@
-
-# if __name__ == '__main__':
-#     unittest.main()
+# pylint: disable-all
```
can we remove this line and have the test pass lint?

kywch replied:
Yap, test_determinism.py now passes linter.

daveey commented on tests/test_deterministic_replay.py:
```diff
@@ -1,120 +1,145 @@
-
-# if __name__ == '__main__':
-#    unittest.main()
+# pylint: disable-all
```
can we remove this?

kywch replied:
done!

daveey commented on tests/testhelpers.py:
```diff
@@ -1,257 +1,209 @@
 # pylint: disable=all
```
can we remove this?

kywch replied:
Yap, testhelpers.py now passes linter.

daveey commented on tests/testhelpers.py:
```diff
@@ -1,257 +1,209 @@
+  cnt_buy = 0
+
+  for entID in list(actions.keys()):
+    for atn, args in actions[entID].items():
```
can this just be a dict rather than a long if statment?

kywch replied:
Got rid of the long if statements.

daveey commented on tests/testhelpers.py:
```diff
@@ -1,257 +1,209 @@
+      assert isinstance(ent.agent, baselines.Scripted) == True, 'All agent must be scripted.'
+
+    # if actions are not provided, generate actions using the scripted policy
+    if actions == {}:
```
could we just use _compute_actions here?

kywch replied:
This part is adapted from the scripted agents' action generation and deserialize in the main Env.step(), which comes after _process_actions. So, the actions generated by scripted agents are not filtered and used as is. 

The part that looks like filtering (`if arg == nmmo.io.action.Price and type(val) != int:`) is to convert discrete values into int and make the actions pickle-able.

daveey reviewed (changes requested):
This looks pretty great. I made a few comments, but I love where this is going. Thanks for fixing a bunch of things!

Main requests:

  - enable lint on the determinism tests
  - add tags support to log_milestone rather than doing it ad-hoc
  - use log.info instead of prints (and add a pylint rule)
  - a few other small requests

kywch committed:
added tags support to realm.log_milestone() and used it

kywch committed:
added a script to run just pylint, xcxc, pytest without touching git

kywch commented:
testhelpers.py, test_determinism.py, test_deterministic_replay.py are on the queue.

kywch committed:
added a pylint check for disallowing print and removed print

kywch commented on nmmo/core/config.py:
```diff
@@ -680,10 +682,10 @@ class Medium(Config):
-  def PLAYERS(self):
-    return [nmmo.Agent]
-
+  PLAYERS = [Agent]
```
This is the line discussed in the discord. Please review.

kywch committed:
fixed the testhelper and determinism test

kywch commented:
Will fix test_deterministic_replay.py tomorrow

kywch committed:
fixed pylint issued in the deterministic replay test

kywch committed:
use pre-git-check.sh to run tests and linter

kywch committed:
fixed ScriptedAgentTestEnv to NOT override step by adding helper functions

kywch committed:
added table() to SerializedState and is_empty() to DataTable

kywch committed:
clear ds tables only in det test but found a weird bug in ItemState -- will fix soon

kywch commented:
This PR got handful already. I'll probably start a new PR for action/item-related commits. Thanks!

daveey commented on nmmo/core/env.py:
```diff
@@ -312,6 +310,24 @@ def _process_actions(self,
+          for arg, val in args.items():
+            atns[atn][arg] = arg.deserialize(self.realm, ent, val)
+            actions[eid] = atns
+        # CHECKME: do we need to remove them from the observations?
```
I don't think we do, since self.obs keeps all the observations, not just neural ones

kywch replied:
Removed!

daveey commented on nmmo/core/realm.py:
```diff
@@ -159,8 +169,16 @@ def step(self, actions):
-    self.log_helper.log_milestone(category, value)
-    self.log_helper.log_event(category, value)
+  def log_milestone(self, category: str, value: float, message: str = None, tags: Dict = None):
+    if self.config.LOG_MILESTONES:
```
Do we need these checks now that we use DummyLoggers?

kywch replied:
These checks are removed. I also saw the same checks in the log_helper.py and agree that log_helper.py is the better place to handle these.

```
  def log_milestone(self, milestone: str, value: float) -> None:
    if self.config.LOG_MILESTONES:
      self._milestone_logger.log(milestone, value)

  def log_event(self, event: str, value: float) -> None:
    if self.config.LOG_EVENTS:
      self._event_logger.log(event, value)
```

daveey commented on tests/testhelpers.py:
```diff
@@ -1,257 +1,200 @@
+
+    return actions
+
+  def _process_actions(self, actions, obs):
```
Can we just assert this in __init__ and not override this method?

kywch replied:
I moved the assert to the __init__ function.

However, the current _process_actions() is not compatible with the output of scripted agents, so it needs to be bypassed for now. I think that's why the original env.step() code had if statements for the scripted agents. 

I'll try to tackle this in the next iteration. I think I'll have to look at both _process_actions() and scripted/baselines.py (and deserialization...)

daveey commented on tests/testhelpers.py:
```diff
@@ -1,257 +1,200 @@
+
+  def _compute_scripted_agent_actions(self, actions):
+    # all agent must be scripted agents when using ScriptedAgentTestEnv
+    for ent in self.realm.players.values():
```
I think we can just do this in __init__

kywch replied:
Yap, moved those to the init.

daveey commented on tests/testhelpers.py:
```diff
@@ -1,257 +1,200 @@
+        # make a copy of actions that are not serialized
+        self.actions[eid] = deepcopy(atns)
+
+        #print(eid, self.actions[eid])
```
remove

kywch replied:
removed!

daveey commented on tests/testhelpers.py:
```diff
@@ -1,257 +1,200 @@
+    for ent in self.realm.players.values():
+      assert isinstance(ent.agent, baselines.Scripted), 'All agent must be scripted.'
+
+    # if actions are not provided, generate actions using the scripted policy
```
I don't understand when we expect actions to be passed in vs not?

kywch replied:
So, the current implementation is very specific to the deterministic replay test using scripted agents. 

The current action output of scripted agents seem to be different from the trained agents, so that's why the current code seems weird. I'll try to tackle this in the next iteration by looking at scripted/baselines.py.

Does this answer your question?

daveey commented on tests/testhelpers.py:
```diff
@@ -1,257 +1,200 @@
+        atns = ent.agent(self.obs[eid])
+
+        # handle the cases that are problematic for pickle
+        for atn, args in atns.items():
```
can we move this into _make_actions_pickleable() or something like that?

daveey replied:
or maybe even _pickle_actions()

kywch replied:
I moved this into test_deterministic_replay.py, where pickling happens, by creating make_actions_picklable()

daveey commented on tests/testhelpers.py:
```diff
@@ -1,257 +1,200 @@
+      for eid in self.actions.keys():
+        assert eid in self.realm.players, f'Entity {eid} not in realm'
+        ent = self.realm.players[eid]
+        atns = deepcopy(self.actions[eid])
```
can we move this into _deserialize_actions()

kywch replied:
yap, moved into env._deserialize_scripted_actions()

kywch committed:
Merge branch 'v1.6' into task-system

kywch committed:
deleted redundant import

pylint error: nmmo/core/config.py:12:0: W0404: Reimport 'Agent' (imported line 9) (reimported)

kywch committed:
refactored testhelpers, etc, incorporating David's suggestions

daveey commented on nmmo/core/env.py:
```diff
@@ -34,6 +34,7 @@ def __init__(self,
     self.obs = None
 
     self.possible_agents = list(range(1, config.PLAYER_N + 1))
+    self.has_scripted_agents = False
```
how about self.scripted_agents = set()

That way we can just loop over those, instead of all the agents

kywch replied:
Changed has_scripted_agents to scripted_agents, which is an ordered set and made necessary changes to use it. 

Thank you so much, David!

daveey reviewed (approved):
One small change request, then feel free to merge. Thanks for making this nice and clean.

kywch committed:
made entity name ent consistent

kywch committed:
changed env.has_scripted_agents to a set one can iterate on
